caster.js

```diff
@@ -0,0 +1,103 @@
+import { EventEmitter } from 'events';
+
+const DEFAULT_IMAGE_URL = '/images/ee_logo_large.png';
+
+const {
+  CastContext,
+  RemotePlayer,
+  RemotePlayerController,
+  RemotePlayerEventType
+} = cast.framework;
+
+const {
+  DEFAULT_MEDIA_RECEIVER_APP_ID,
+  GenericMediaMetadata,
+  LoadRequest,
+  MediaInfo,
+  MetadataType,
+  StreamType
+} = chrome.cast.media;
+
+const { AutoJoinPolicy, Image } = chrome.cast;
+
+export default class Caster extends EventEmitter {
+  constructor() {
+    super();
+    this.remotePlayer = null;
+    this.remotePlayerController = null;
+    this.mediaSource = null;
+    this.initInterval = null;
+    this.isPlaying = false;
+
+    this.onConnectedChanged = this.switchPlayer.bind(this);
+    this.onPausedChanged = this.togglePause.bind(this);
+  }
+
+  get castContext() {
+    return CastContext.getInstance();
+  }
+
+  get isConnected() {
+    return this.remotePlayer && this.remotePlayer.isConnected;
+  }
+
+  init(options = {}) {
+    // TODO: Replace with receiver application ID when created in the Google Cast Developer Console
+    options.receiverApplicationId = DEFAULT_MEDIA_RECEIVER_APP_ID;
+    // Auto join policy can be one of the following three:
+    // ORIGIN_SCOPED - Auto connect from same appId and page origin
+    // TAB_AND_ORIGIN_SCOPED - Auto connect from same appId, page origin, and tab
+    // PAGE_SCOPED - No auto connect
+    options.autoJoinPolicy = AutoJoinPolicy.ORIGIN_SCOPED;
+
+    this.castContext.setOptions(options);
+
+    this.remotePlayer = new RemotePlayer();
+    this.remotePlayerController = new RemotePlayerController(this.remotePlayer);
+    this.remotePlayerController.addEventListener(
+      RemotePlayerEventType.IS_CONNECTED_CHANGED,
+      this.onConnectedChanged
+    );
+    this.remotePlayerController.addEventListener(
+      RemotePlayerEventType.IS_PAUSED_CHANGED,
+      this.onPausedChanged
+    );
+  }
+
+  switchPlayer() {
+    if (this.remotePlayer.isConnected) return this.emit('cast:start');
+    return this.emit('cast:stop');
+  }
+
+  togglePause(isPaused) {
+    this.isPlaying = !isPaused.value;
+  }
+
+  play() {
+    const mediaInfo = new MediaInfo(this.mediaSource.location, 'audio/ogg');
```

Comment:
Hardcoding `audio/ogg`? 🤔

Reply:
Yeah, missed that one during refactor.

Though not sure what's the best way to handle this; ideally, MIME type would be inferred from the codec data, but that might be a bit involved to implement.

There are quite a few `audio` MIME types: https://www.iana.org/assignments/media-types/media-types.xhtml#audio
So we would need a way to do some kind of mapping between codec data -> MIME type, and default to `audio/ogg` or something? Though in the end, how important is it really for the MIME type to be 100% correct? I think browsers will handle them relatively the same, as long as it's an `audio/*` that the browser supports: https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats#Browser_compatibility.

What are your thoughts?

Reply:
Agree with everything you said. Tested with wildcard mimetype `audio/*` and it worked flawlessly 🎉